### ISSUE TITLE
I can't understand the code in eval()

### ISSUE THREAD

**Issue by magnificGH** [Kyubyong/transformer#121]
![图片](https://user-images.githubusercontent.com/26301211/59574986-11fd3800-90ec-11e9-9eb9-eac3c249192c.png)

y_hat should be [N, t], and tf.reduce_sum(y_hat, 1) should be [N,], I can't understand this if statement. And decoder_inputs should be [N, t], i think tf.concat((decoder_inputs, y_hat), 1) is wrong, should it be changed to tf.concat((decoder_inputs, y_hat[:,-1]), 1)

**Comment by WalesExcitedMei** [comment]
Have you figured it out? I do not know the meaning either.

**Comment by qwtao** [comment]
me too

**Comment by yichaopku** [comment]
i think there are two problems in the code. the first one is mask used to set only the previous context visible. the second one is the problems methioned in this issue.

**Comment by yichaopku** [comment]
I think there must be something wrong with this piece of codes. The if statement have effect only when constructing the calculation graph. So i think the if statement could be ignored. If we want the decoding length not to be fixed, we'd better use the while_loop operator in tensorflow. As for the line  tf.concat((decoder_inputs, y_hat), 1), the decoder_inputs always has shape [N, 1] and filled with the id of <s>, so it is not wrong.

**Comment by zhongbin1** [comment]
I also think the if statement in first line has something wrong, because the condition will never be satisfied.

**Comment by gaoxuezhao** [comment]
> I also think the if statement in first line has something wrong, because the condition will never be satisfied.

I cannot agree more.

**Comment by BetterZH** [comment]
I think the if statement in first line is wrong.And my code：
last_words=tf.reshape(y_hat[:,1],[-1,1])                                                      # [N,1]
if tf.reduce_sum(last_words, 0) == self.token2idx["<pad>"]: break        #  [1]  if  sum=0

**Comment by bozhenhhu** [comment]
> I think the if statement in first line is wrong.And my code：
> last_words=tf.reshape(y_hat[:,1],[-1,1]) # [N,1]
> if tf.reduce_sum(last_words, 0) == self.token2idx[""]: break # [1] if sum=0
May be it shoud be:
            if tf.reduce_sum( tf.reshape(y_hat[:, 0], [-1, 1]), 0) == self.token2idx["<pad>"]: break

**Comment by LiChao-cy** [comment]
Is there some mistake in the evaluation codes? @Kyubyong
